nifi-nar-bundles/nifi-framework-bundle/nifi-framework/nifi-web/nifi-web-api/src/main/java/org/apache/nifi/web/api/ProcessGroupResource.java

```diff
@@ -325,6 +441,859 @@ public Response updateProcessGroup(
+            throw new IllegalArgumentException("Revision must be specified.");
+        }
+
+        // In order to update variables in a variable registry, we have to perform the following steps:
+        // 1. Determine Affected Components (this includes any Processors and Controller Services and any components that reference an affected Controller Service).
+        //    1a. Determine ID's of components
+        //    1b. Determine Revision's of associated components
+        // 2. Stop All Affected Processors
+        // 3. Disable All Affected Controller Services
+        // 4. Update the Variables
+        // 5. Re-Enable all Affected Controller Services (services only, not dependent components)
+        // 6. Re-Enable all Processors that Depended on the Controller Services
+
+        // Determine the affected components (and their associated revisions)
+        final VariableRegistryEntity computedEntity = serviceFacade.populateAffectedComponents(requestEntity.getVariableRegistry());
+        final VariableRegistryDTO computedRegistryDto = computedEntity.getVariableRegistry();
+        if (computedRegistryDto == null) {
+            throw new ResourceNotFoundException(String.format("Unable to locate group with id '%s'.", groupId));
+        }
+
+        final Set<AffectedComponentDTO> affectedComponents = serviceFacade.identifyComponentsAffectedByVariableRegistryUpdate(requestEntity.getVariableRegistry());
+
+        final Map<String, List<AffectedComponentDTO>> affectedComponentsByType = affectedComponents.stream()
+            .collect(Collectors.groupingBy(comp -> comp.getComponentType()));
+
+        final List<AffectedComponentDTO> affectedProcessors = affectedComponentsByType.get(AffectedComponentDTO.COMPONENT_TYPE_PROCESSOR);
+        final List<AffectedComponentDTO> affectedServices = affectedComponentsByType.get(AffectedComponentDTO.COMPONENT_TYPE_CONTROLLER_SERVICE);
+
+
+        if (isReplicateRequest()) {
+            // update the variable registry
+            final VariableRegistryUpdateRequest updateRequest = createVariableRegistryUpdateRequest(groupId);
+            updateRequest.getIdentifyRelevantComponentsStep().setComplete(true);
+            final URI originalUri = getAbsolutePath();
+
+            // Submit the task to be run in the background
+            final Runnable taskWrapper = () -> {
+                try {
+                    updateVariableRegistryReplicated(groupId, originalUri, affectedProcessors, affectedServices, updateRequest, requestEntity);
+                } catch (final Exception e) {
+                    logger.error("Failed to update variable registry", e);
+                    updateRequest.setFailureReason("An unexpected error has occurred: " + e);
+                }
+            };
+
+            variableRegistryThreadPool.submit(taskWrapper);
+
+            final VariableRegistryUpdateRequestEntity responseEntity = new VariableRegistryUpdateRequestEntity();
```

Comment:
Looking at this more, I think the entity class should really just extend Entity, not ComponentEntity. In such a case I dont think we need the permissions DTO on the entity.